### ISSUE TITLE
Support for adding an EndpointRequest request matcher for the actuator base path

### ISSUE THREAD

**Issue by filiphr** [spring-projects/spring-boot#13183]
We are adding multiple Spring Security configurations and we want to have the actuator protected in it's own configuration. We are doing the following:

```java
http
 .requestMatcher(EndpointRequest.toAnyEndpoint())
 .authorizeRequests()
 .requestMatchers(EndpointRequest.to(InfoEndpoint.class, HealthEndpoint.class)).authenticated()
 .requestMatchers(EndpointRequest.toAnyEndpoint()).hasAnyAuthority(DefaultPrivileges.ACCESS_ADMIN)
  .and().httpBasic();
```

If I do `EndpointRequest.toLinks()` are used then only the `/actuator` is matched and not `/actuator/**`

**Comment by mbhave** [comment]
`EndpointRequest.toAnyEndpoint()` includes all actuators including the links (unless you explicitly exclude links). Could you explain a bit more why `.requestMatcher(EndpointRequest.toAnyEndpoint())` doesn't do what you want for restricting this configuration just to the actuator endpoints?

**Comment by filiphr** [comment]
It does, but in the moment that creates N `OrRequestMatcher`(s) (for each endpoint one) instead of only one (`/actuator/**`). Which means that on each request each endpoint would need to go through each of them in order to determine if this security filter chain should be called. See [here](https://github.com/spring-projects/spring-security/blob/master/web/src/main/java/org/springframework/security/web/FilterChainProxy.java#L226).

I would like to avoid prefer to have a single `RequestMatcher` if possible.

From what I understood (from my debugging today). For each `WebSecurityConfigurerAdapter` a single `FilterSecurityChain` will be created.

**Comment by filiphr** [comment]
Thanks.

In theory you don't have to add a new method. You can just adapt the [delegate creation](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/security/servlet/EndpointRequest.java#L204) in the to create `pathMappedEndpoints.getBasePath() + "/**"` when `includes` and `excludes` are empty and there is a base path and links are included. Which means when one uses `EndpointRequest.toAnyEndpoint()` then by default `/{base-path}/**` would be used (which is what I expected would happen when I used that).

This is just from the top of my head in this moment

**Comment by filiphr** [comment]
My idea was that if no base path is set then it will do the same as now

**Comment by philwebb** [comment]
> Well it already does that for the links.

It doesn't match any suffix for the links, there's no `/**` suffix.

What's the main reason that you want a `/**` suffix? I'm still not sure I totally understand what additional paths you're using that aren't covered by `toAnyEndpoint()`?

**Comment by filiphr** [comment]
I was referring to:

>I don't like the idea of `toAnyEndpoint` behaving differently depending on if the `basePath` is set.

The difference and the reason for this issue it to add the `/**` suffix either through the current `toAnyEndpoint` or a new method as a helper method.

The reason is when there are multiple security configurations, one for actuator, and then some other then Spring Security created multiple `SecurityFilterChain`(s). Then on each request in the [`FilterChainProxy`](https://github.com/spring-projects/spring-security/blob/master/web/src/main/java/org/springframework/security/web/FilterChainProxy.java#L224) each `SecurityFilterChain` is checked if it is matched. This means if the actuators are first then for each request all endpoints request matchers will be checked if they matched as `toAnyEndpoint()` creates an `OrRequestMatcher` with all endpoints. 

The problem is not in the paths, with a suffix or `toAnyEndpoint()` it works the same. The only difference is the amount of `RequestMatcher`(s) that would be evaluated on each request.

**Comment by philwebb** [comment]
> The reason is when there are multiple security configurations, one for actuator, and then some other then Spring Security created multiple SecurityFilterChains

Is this actually causing you a problem? Are you finding you have performance issues? Or does it make debugging harder? What's the underlying reason why we shouldn't create multiple matchers?

**Comment by filiphr** [comment]
I am going to check tomorrow in regard of performance issues just to be sure and that I can show some numbers here.

As for debugging yes it makes it a bit more difficult as there are more matchers in the logs. However, I can live with it.

**Comment by filiphr** [comment]
I created a [demo project](https://github.com/filiphr/demo-actuator-security-performance) where one can pick between using a base path `AntMatcher` or the `EndpointRequest#toAnyEndpoint()`.

I did some runs with 100000 iterations calling the greeting controller and I get
* 4175 (3963) req/s when base path matcher is used
* 3691 (3157) req/s when `EndpointReques#toAnyEndpoint()` is used

The numbers are computed by invoking the actuator metrics endpoint.

Feel free to play around with the project. Both numbers are of course good and in a real world scenario they might defer, but I just wanted to show that it might make a difference.
